gemstone/src/swapper/stargate/provider.rs

```diff
@@ -0,0 +1,307 @@
+        endpoint
+            .pools
+            .iter()
+            .find(|x| x.asset.id == *asset)
+            .map(|x| x.address.clone())
+            .ok_or(SwapperError::NotSupportedChain)
+    }
+
+    pub fn get_asset_decimals(&self, asset: &AssetId) -> Result<i32, SwapperError> {
+        let endpoint = self.enpoints.iter().find(|x| x.id == asset.chain).ok_or(SwapperError::NotSupportedChain)?;
+        match endpoint.pools.iter().find(|x| x.asset.id == *asset) {
+            Some(pool) => Ok(pool.asset.decimals),
+            None => Err(SwapperError::NotSupportedChain),
+        }
+    }
+}
+
+#[async_trait]
+impl GemSwapProvider for Stargate {
+    fn provider(&self) -> SwapProvider {
+        SwapProvider::Stargate
+    }
+
+    fn supported_assets(&self) -> Vec<SwapChainAsset> {
+        self.enpoints
+            .iter()
+            .map(|x| SwapChainAsset::Assets(x.id, x.pools.iter().map(|y| y.asset.id.clone()).collect()))
+            .collect()
+    }
+
+    async fn fetch_quote(&self, request: &SwapQuoteRequest, provider: Arc<dyn AlienProvider>) -> Result<SwapQuote, SwapperError> {
+        let from_asset = &request.from_asset;
+        let to_asset = &request.to_asset;
+
+        if from_asset.chain == to_asset.chain {
+            return Err(SwapperError::NotSupportedChain);
+        }
+
+        if from_asset.is_native() && !to_asset.is_native() {
+            return Err(SwapperError::NotSupportedChain);
+        }
+
+        let amount_ld = U256::from_str(request.value.as_str()).unwrap();
+
+        let pool = self.get_pool(from_asset).unwrap();
+
+        let mut send_param = SendParam {
+            dstEid: self.get_endpoint_id(&to_asset.chain).unwrap(),
+            to: self.address_to_bytes32(request.destination_address.as_str()),
+            amountLD: amount_ld,
+            minAmountLD: amount_ld,
+            extraOptions: Bytes::from_str("0x").unwrap(),
+            composeMsg: Bytes::from_str("0x").unwrap(),
+            oftCmd: Bytes::from_str("0x").unwrap(),
+        };
+
+        // Encode call data
+        let call_data = IStargate::quoteOFTCall {
+            _sendParam: send_param.clone(),
+        }
+        .abi_encode();
+
+        let call = EthereumRpc::Call(TransactionObject::new_call(pool.as_str(), call_data), BlockParameter::Latest);
+        let response: JsonRpcResult<String> = jsonrpc_call(&call, provider.clone(), &request.from_asset.chain).await?;
+        let result = response.take()?;
+        let hex_data = hex::decode(result).map_err(|e| SwapperError::NetworkError { msg: e.to_string() })?;
+        let quote_oft_data = IStargate::quoteOFTCall::abi_decode_returns(&hex_data, true).map_err(|e| SwapperError::ABIError { msg: e.to_string() })?;
+
+        send_param.minAmountLD = apply_slippage_in_bp(&quote_oft_data.receipt.amountReceivedLD, request.options.slippage_bps);
+
+        let messaging_fee_calldata = IStargate::quoteSendCall {
```

Comment:
take a look at across/hubpool.rs, we prefer implementing contract client for the purpose of reuse later